lib/ukcpuid/cpuid.c

```diff
@@ -0,0 +1,182 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include <uk/errptr.h>
+#include <uk/cpuid.h>
+
+#include "../plat/common/include/x86/cpu_defs.h"
+
+static struct uk_cpuid_info info;
+static struct uk_cpuid_cache cache;
+
+static unsigned int __highest_basic_func_value;
+static unsigned int __highest_extended_func_value;
+static unsigned int last_eax_function = -1;
+static unsigned int last_ecx_function = -1;
+
+
+unsigned char is_flag_enabled(unsigned int eax_function, unsigned int ecx_function,
+							  unsigned int flag, unsigned int *reg)
+{
+
+	if (last_eax_function != eax_function ||
+		last_ecx_function != ecx_function) {
+		last_eax_function = info.EAX = eax_function;
+		last_ecx_function = info.ECX = ecx_function;
+
+		cpuid_macro(info.EAX, info.EBX, info.ECX, info.EDX, info.EAX, info.ECX);
+	}
+
+	return (*reg) & flag;
+}
+
+void init_cpuid(void)
+{
+	info.EAX = EAX_HIGHEST_BASIC_VALUE;
+	info.ECX = 0x0;
+
+	cpuid_macro(info.EAX, info.EBX, info.ECX, info.EDX, info.EAX, info.ECX);
+	__highest_basic_func_value = info.EAX;
+
+	info.EAX = EAX_HIGHEST_EXTENDED_VALUE;
+	cpuid_macro(info.EAX, info.EBX, info.ECX, info.EDX, info.EAX, info.ECX);
+	__highest_extended_func_value = info.EAX;
+}
+
+unsigned char is_xsave_available(void)
+{
+	if (__highest_basic_func_value < EAX_FEATURE_INFO) {
+		return 0;
+	}
+
+	if (cache.is_xsave_cached) {
+		goto return_from_cache;
+	}
+
+	cache.is_xsave_cached = 1;
+	cache.is_xsave_available = is_flag_enabled(EAX_FEATURE_INFO, 0x0, X86_CPUID1_ECX_XSAVE, &(info.ECX));
```

Comment:
I wonder if it makes sense to cache a whole response instead since I expect there are multiple bits we could query, right?

Reply:
Yep, could be an option. Another option it would be to store multiple bits (not all) when a response is available. E.g: when cpuid is called with EAX=0, ECX=0 we could store bits 1, 3, 12..etc. I don't know how much sense it makes, though. What is your view on that?

Reply:
@samoilescusebastian , from my point of view, each query you make should store / cache all relevant information in the response, irrespective of whether you intended to get all that information or only part of it. It doesn't make sense, from my point of view, to get, say, 4 pieces of information when doing a query and only store 1.